src/compiler/checker.rs

```diff
@@ -0,0 +1,363 @@
+        for type_ in types {
+            self.add_type_to_union(type_set, type_.clone());
+        }
+    }
+
+    fn get_union_type(&self, types: Vec<Rc<Type>>) -> Rc<Type> {
+        let mut type_set: Vec<Rc<Type>> = vec![];
+        self.add_types_to_union(&mut type_set, /*TypeFlags::empty(), */ &types);
+        self.get_union_type_from_sorted_list(type_set)
+    }
+
+    fn get_union_type_from_sorted_list(&self, types: Vec<Rc<Type>>) -> Rc<Type> {
+        let mut type_: Option<Rc<Type>> = None;
+        if type_.is_none() {
+            let base_type = self.create_type(TypeFlags::Union);
+            type_ = Some(Rc::new(
+                (UnionType {
+                    _union_or_intersection_type: BaseUnionOrIntersectionType {
+                        _type: base_type,
+                        types,
+                    },
+                })
+                .into(),
+            ));
+        }
+        type_.unwrap()
+    }
+
+    fn is_fresh_literal_type(&self, type_: Rc<Type>) -> bool {
+        if !type_.flags().intersects(TypeFlags::Literal) {
+            return false;
+        }
+        match &*type_ {
+            Type::IntrinsicType(intrinsic_type) => match intrinsic_type {
+                IntrinsicType::FreshableIntrinsicType(freshable_intrinsic_type) => {
+                    ptr::eq(&*type_, freshable_intrinsic_type.fresh_type().as_ptr())
+                }
+                _ => panic!("Expected FreshableIntrinsicType"),
+            },
+            _ => panic!("Expected IntrinsicType"),
+        }
+    }
+
+    fn is_type_assignable_to(&self, source: Rc<Type>, target: Rc<Type>) -> bool {
+        self.is_type_related_to(source, target, &self.assignable_relation)
+    }
+
+    fn is_type_related_to(
+        &self,
+        mut source: Rc<Type>,
+        target: Rc<Type>,
+        relation: &HashMap<String, RelationComparisonResult>,
+    ) -> bool {
+        if self.is_fresh_literal_type(source.clone()) {
+            source = match &*source {
+                Type::IntrinsicType(intrinsic_type) => match intrinsic_type {
+                    IntrinsicType::FreshableIntrinsicType(freshable_intrinsic_type) => {
+                        freshable_intrinsic_type.regular_type().upgrade().unwrap()
+                    }
+                    _ => panic!("Expected IntrinsicType"),
+                },
+                _ => panic!("Expected IntrinsicType"),
+            };
+        }
+        if source
+            .flags()
+            .intersects(TypeFlags::StructuredOrInstantiable)
+            || target
+                .flags()
+                .intersects(TypeFlags::StructuredOrInstantiable)
+        {
+            return self.check_type_related_to(source, target, relation);
+        }
+        false
+    }
+
+    fn check_type_related_to(
+        &self,
+        source: Rc<Type>,
+        target: Rc<Type>,
+        relation: &HashMap<String, RelationComparisonResult>,
+    ) -> bool {
+        false
```

Comment:
This is what's actually triggering the type error on the `true` operand in `++true`. So I'd have to implement real "is `source` assignable to `target`?" logic here in order for eg `++1` to not get flagged but `++true` to get flagged